### PR TITLE
Add Chip Design environment

### PR DESCRIPTION
**Pull request by younik**
This PR introduces a new discrete GFlowNet environment for chip macro placement (`ChipDesign`), along with supporting helper code, example training scripts, and tests.

## Changes Made

- **New environment**: `ChipDesign` environment + `ChipDesignStates`, backed by Circuit Training's `plc_wrapper_main` via a local client wrapper.
- **Helper package**: Chip-design helper package with test netlists/placements, container/binary setup scripts, and container build helpers.
- **Training scripts & tests**: Tutorial training scripts, smoke tests, and environment tests.
- **Bug fix**: `get_blockages_from_comments` in `utils.py` now normalizes a single `str` input to `[str]` before iterating, preventing character-by-character iteration when a single filename is passed.
- **Resource management**: Added `close()` and `__del__()` methods to `ChipDesign` to properly shut down the `PlacementCost` subprocess and prevent resource leaks across tests or repeated runs.

# Updates (April 2026)

## Merged with master

  - Resolved all conflicts with latest master (583 commits)
  - Adapted ChipDesign env to master's lazy mask caching API (DiscreteStates property-based masks, debug flag, conditions/device/debug kwargs)

## Singularity/Apptainer support

  - singularity_image parameter on PlacementCost, ChipDesign, and training CLI (--singularity_image auto)
  - _resolve_plc_binary() auto-detection: env var > native binary > apptainer wrapper
  - build_container.py and plc_wrapper.def for building .sif images
  - Compatible with both Singularity and Apptainer

## Coordinate descent orientation optimization

  - optimize_orientations() sweeps all 8 DEF/LEF orientations per macro via greedy coordinate descent (matching Google's cd_finetune approach)
  - Called in log_reward after analytical placement, controlled by cd_finetune=True (default)
  - Consistently improves placement quality (~3-5% reward gain)

## Reward normalization

  - reward_norm parameter: None (raw), "zscore", or "minmax" — estimated from random placement samples at init
  - reward_temper for temperature scaling (higher = sharper distribution)
  - CostStats.precompute() for batch pre-computation across multiple netlists — intended workflow for conditional GFlowNets
  - Without normalization reward ratio is ~1.3x; with zscore ~33x; with minmax ~2.7x

## Bug fixes
  - Fixed CUDA/CPU device mismatch in States._compare (.to() is no-op on same device, torch.compile safe)
  - Fixed sink state handling in update_masks
  - Fixed _apply_state_to_plc sentinel guard (loc >= 0 instead of loc != -1)
  - Fixed duplicate [tool.pytest] / [tool.pytest.ini_options] in pyproject.toml
  - Added graceful FileNotFoundError skip in tests when binary unavailable

### CODE REVIEW

File: src/gfn/gym/helpers/chip_design/utils.py [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+        w, h = plc.get_node_width_height(idx)
+        return w * h
+
+    canvas_width, canvas_height = plc.get_canvas_width_height()
+
+    def distance_to_edge(idx):
+        x, y = plc.get_node_location(idx)
+        return min(
+            x, y, canvas_width - x - canvas_width, canvas_height - y - canvas_height
+        )
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/gym/helpers/chip_design/utils.py]: @younik I agree with copilot here - what was the logic behind this math?

File: pyproject.toml [hunk 2]
```diff
@@ -176,6 +176,23 @@ reportUnboundVariable = "warning"
...
+filterwarnings = [
+    "ignore::pytest.PytestConfigWarning",
+]
+reportUnusedImport = true
+reportUnusedVariable = true
+reportOptionalSubscript = true
+reportOptionalCall = true
+reportOptionalIterable = true
+reportOptionalContextManager = true
+reportOptionalOperand = true
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply pyproject.toml]: @copilot can you make a new PR that only changes `pyproject.toml` and submits it to `master`? This is a distinct issue that does not belong in this pr.
